### ISSUE TITLE
support multi-pass feature encoders

### ISSUE THREAD

**Issue by bethard** [ClearTK/cleartk#257]
Original [issue 259](https://code.google.com/p/cleartk/issues/detail?id=259) created by ClearTK on 2011-10-20T16:53:21.000Z:

I would like to be able to declare feature extractors like:
    new ZeroMeanOneStddevNormalized(featureExtractor);
and have all the features extracted by featureExtractor normalized using the mean and standard deviation of the feature values in the training set.

Similarly, I think we should be able to write things like:
    new InverseDocumentFrequencyNormalized(wordCountsExtractor);
and have all the features extracted by the wordCountsExtractor normalized using the inverse document frequency of the feature values in the training set.

What both of these have in common is that they have to take two passes through the data: one where global statistics are collected and an initial version of the features is created, and one where the initial version of the features are modified based on the collected global statistics.

I think we could make this work by allowing multi-pass feature encoders. Data writers would run as they currently do, but they would write out a serialization of the Instances alongside the training data. If a feature encoder required a second pass, e.g. if it was tagged with an interface like MultiPassFeatureEncoder, then the data writer would detect this, and run a second pass where it loaded the serialized Instances, and passed them through the feature encoder a second time, rewriting the training data.
